crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1665 @@
+
+            self.row_group_data_receivers = row_group_data_receivers;
+            self.row_group_data_task_handle = Some(row_group_data_task_handle);
+
+            let row_group_decoder = self.init_row_group_decoder();
+            self.row_group_decoder = Some(Arc::new(row_group_decoder));
+        }
+
+        fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) {
+            use std::sync::atomic::Ordering;
+
+            assert!(recv.is_empty());
+            assert_eq!(send.len(), 1);
+
+            if self.is_finished.load(Ordering::Relaxed) {
+                send[0] = PortState::Done;
+                assert!(
+                    self.row_group_data_task_handle.is_none(),
+                    "should have already been shut down"
+                );
+            } else if send[0] == PortState::Done {
+                {
+                    // Early shutdown - our port state was set to `Done` by the downstream nodes. This
+                    // also means we can discard any potential errors from the `shutdown()`. Note that
+                    // `shutdown()` internally unwraps any critical errors.
+                    let _ = polars_io::pl_async::get_runtime().block_on(self.shutdown());
+                };
+                self.is_finished.store(true, Ordering::Relaxed);
+            } else {
+                send[0] = PortState::Ready
+            }
+        }
+
+        fn spawn<'env, 's>(
+            &'env mut self,
+            scope: &'s TaskScope<'s, 'env>,
+            recv: &mut [Option<RecvPort<'_>>],
+            send: &mut [Option<SendPort<'_>>],
+            _state: &'s ExecutionState,
+            join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+        ) {
+            use std::sync::atomic::Ordering;
+
+            assert!(recv.is_empty());
+            assert_eq!(send.len(), 1);
+
+            let morsel_senders = send[0].take().unwrap().parallel();
+
+            // Note: We could have less receivers than `morsel_senders`, if the row group prefetch
+            // size is small relative to `num_pipelines`.
+            assert!(self.row_group_data_receivers.len() <= morsel_senders.len());
+            assert!(self.spawned_task_handles.is_empty());
+            assert!(self.spawned_task_handles.capacity() >= self.row_group_data_receivers.len());
+            assert!(!self.is_finished.load(Ordering::Relaxed));
+
+            let row_group_decoder = self.row_group_decoder.as_ref().unwrap().clone();
+
+            if let Some(v) = self.row_group_data_task_starter.take() {
+                v.send(()).unwrap();
+            }
+
+            for (mut row_group_data_rx, mut morsel_tx) in
+                self.row_group_data_receivers.drain(..).zip(morsel_senders)
+            {
+                let row_group_decoder = row_group_decoder.clone();
+                let is_finished = self.is_finished.clone();
+
+                self.spawned_task_handles
```

Comment:
It's for amortizing the vec allocation across execution phases. Although it really is a micro-optimization 😄, if you'd like I can just make it a local vec to avoid confusion?

Reply:
Yes, there are only a couple phases, really not perf sensitive, please just make that a local variable.

Reply:
Updated